pkg/cli/dirs.go

```diff
@@ -0,0 +1,68 @@
+// Copyright © 2021 The Tekton Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func CacheDir() string {
+	tmpDir, err := os.UserCacheDir()
+	if err != nil {
+		tmpDir = os.TempDir()
+	}
+	tmpDir, err = contractHome(tmpDir)
+	if err != nil {
+		tmpDir = os.TempDir()
+	}
+	return filepath.Join(tmpDir, "tkn")
+}
+
+func ConfigDir() (string, error) {
+	configDir, err := os.UserConfigDir()
+	if err != nil {
+		return "", err
+	}
+	configDir, err = contractHome(configDir)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(configDir, "tkn"), err
+}
+
+// opposite of homedir.Expand()
+func contractHome(path string) (string, error) {
+	if len(path) == 0 {
+		return path, nil
+	}
+
+	if path[0] == '~' {
+		return path, nil
+	}
+
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	if !strings.HasPrefix(path, homeDir) {
+		return "", errors.New("cannot contract user-specific home dir")
+	}
+
+	return filepath.Join("~", path[len(homeDir):]), nil
+}
```

Comment:
How does this work / act on a Windows machine ? 🤔

Reply:
Good question. I'll investigate and add some tests

Reply:
https://github.com/tektoncd/cli/blob/0c342d0a35b01d9e5bf0ad12e66a76a175541447/pkg/bundle/flags.go#L67

I tried a couple things:

`flags.StringVar(&opts.cacheDir, "cache-dir", "ping/pong", "A directory to cache Tekton bundles in.")` shows ping/pong fine
`flags.StringVar(&opts.cacheDir, "cache-dir", "ping\\pong", "A directory to cache Tekton bundles in.")` (in quotes `\\` is to have a literal `\`) shows ping\\pong so I guess this is a cobra issue??

Reply:
I think that's good enough for me for windows support, we need some proper testing and support on windows which we don't have currently...

Reply:
Thanks @06kellyjac

Reply:
I've opened an issue in the cobra project to ask about the doubled up backslashes